File: semestar_6/ooup/lab/lab3/task1/parrot.c
#include <stdlib.h>

typedef const char* (*PTRFUN)();
typedef struct Parrot_Vtable Parrot_Vtable;
typedef struct Parrot Parrot;

struct Parrot_Vtable {
	PTRFUN name;
    PTRFUN greet;
    PTRFUN menu;
};

struct Parrot {
    const Parrot_Vtable *vtbl;
    const char *name;
};

const char *name_parrot_impl(const Parrot *parrot) {
	return parrot->name;
}

const char *greet_parrot_impl(void) {
	return "papiga!";
}

const char *menu_parrot_impl(void) {
	return "bilo sto";
}

static Parrot_Vtable vtbl = {
	.name = &name_parrot_impl,
	.greet = &greet_parrot_impl,
	.menu = &menu_parrot_impl
};

size_t sizeOf(void) {
	return sizeof(Parrot);
}

void construct(void *obj, const char *name) {
	Parrot *p = (Parrot *) obj;
	p->vtbl = &vtbl;
	p->name = name;
}

void *create(const char *name) {
	void *obj = malloc(sizeOf());
	construct(obj, name);
	return obj;
}

File: semestar_6/ooup/lab/lab3/task12/src/main/java/hr/fer/zemris/ooup/lab3/model/plugins/Parrot.java
package hr.fer.zemris.ooup.lab3.model.plugins;

import hr.fer.zemris.ooup.lab3.model.NamedAnimal;

public class Parrot extends NamedAnimal {

	public Parrot(String name) {
		super(name);
	}

	@Override
	public String greet() {
		return "papiga!";
	}

	@Override
	public String menu() {
		return "bilo što";
	}

	

}
